File: fast_sync_typescript_service_sdk/packages/fast-sync-core/src/core/implementation/service/conflicts_handler.ts
import { ISyncableObject } from "../../abstraction/metadata/ISyncable_object";
import { IConflictsHandler } from "../../abstraction/service/IConflicts_handler";
import { ConflictsResolutionStrategyEnum } from "../../abstraction/service/conflicts_resolution_strategie";
import { isNullOrUndefined } from "../utils/utils";

export class ConflictsHandler implements IConflictsHandler {

    constructor(private resolutionStrategy: ConflictsResolutionStrategyEnum = ConflictsResolutionStrategyEnum.LastWriterWins,
        private conflictsResolutionFunction?: (oblObject: ISyncableObject, newObject: ISyncableObject) => Promise<ISyncableObject>) {

        if (ConflictsResolutionStrategyEnum.PredefinedRules == this.resolutionStrategy && isNullOrUndefined(conflictsResolutionFunction)) {
            throw new Error("conflictsResolutionFunction is required when you put ConflictsResolutionStrategyEnum.PredefinedRules");
        }
    }

    public getConflictsResolutionStrategy() {
        return this.resolutionStrategy;
    }

    public async resolveConflicts(serverObject: ISyncableObject, clientObject: ISyncableObject): Promise<ISyncableObject> {
        if (ConflictsResolutionStrategyEnum.LastWriterWins == this.resolutionStrategy) {
            return clientObject;
        }

        if (ConflictsResolutionStrategyEnum.TimestampOrdering == this.resolutionStrategy) {
            return serverObject.metadata.timestamp > clientObject.metadata.timestamp ? serverObject : clientObject;
        }

        if (ConflictsResolutionStrategyEnum.PredefinedRules == this.resolutionStrategy) {
            // check if there is other then client increment, ==> some one else did change the objct
            // ==> the client is late ==> conflict
            let clientVersion: number = clientObject.metadata.version;
            let serverVersion: number = serverObject.metadata.version;
            if (Math.abs(clientVersion - serverVersion) != 1 && serverVersion >= clientVersion) {
                let resolvedobject = await this.conflictsResolutionFunction(serverObject, clientObject);
                resolvedobject.metadata = clientObject.metadata;
                return resolvedobject
            }
            return clientObject;

        }
    }
}

File: fast_sync_dotnet_service/fast-sync-core/implementation/service/ConflictsHandler.cs
using fast_sync_core.abstraction.data;
using fast_sync_core.implementation.data;

namespace fast_sync_core.implementation
{
    public class ConflictsHandler : IConflictsHandler
    {
        private readonly ConflictsResolutionStrategyEnum _resolutionStrategy;
        private readonly Func<SyncableObject, SyncableObject, Task<SyncableObject>>? _conflictsResolutionFunction;

        public ConflictsHandler(
            Func<SyncableObject, SyncableObject, Task<SyncableObject>> conflictsResolutionFunction,
            ConflictsResolutionStrategyEnum resolutionStrategy = ConflictsResolutionStrategyEnum.LastWriterWins 
            )
        {
            _resolutionStrategy = resolutionStrategy;
            _conflictsResolutionFunction = conflictsResolutionFunction;

            if (_resolutionStrategy == ConflictsResolutionStrategyEnum.PredefinedRules && _conflictsResolutionFunction == null)
            {
                throw new ArgumentException("conflictsResolutionFunction is required when you put ConflictsResolutionStrategyEnum.PredefinedRules");
            }
        }

        public ConflictsResolutionStrategyEnum GetConflictsResolutionStrategy()
        {
            return _resolutionStrategy;
        }

        public async Task<T> ResolveConflicts<T>(T serverObject, T clientObject) where T : SyncableObject
        {
            if (_resolutionStrategy == ConflictsResolutionStrategyEnum.LastWriterWins)
            {
                return clientObject;
            }

            if (_resolutionStrategy == ConflictsResolutionStrategyEnum.TimestampOrdering)
            {
                return serverObject.Metadata.Timestamp > clientObject.Metadata.Timestamp ? serverObject : clientObject;
            }

            if (_resolutionStrategy == ConflictsResolutionStrategyEnum.PredefinedRules)
            {
                if(_conflictsResolutionFunction == null)
                {
                    throw new ArgumentNullException("conflictsResolutionFunction is null for the type" + typeof(T).Name); 
                }
                int clientVersion = clientObject.Metadata.Version;
                int serverVersion = serverObject.Metadata.Version;
                if (Math.Abs(clientVersion - serverVersion) != 1 && serverVersion >= clientVersion)
                {
                    SyncableObject resolvedObject = await _conflictsResolutionFunction.Invoke(serverObject, clientObject);
                    resolvedObject.Metadata = clientObject.Metadata;
                    return (T)resolvedObject;
                } 
                return clientObject;
            }

            return clientObject;
        }
    }
}
